mlflow/transformers/llm_inference_utils.py

```diff
@@ -0,0 +1,127 @@
+_METADATA_LLM_INFERENCE_TASK_KEY = "task"
+
+_LLM_INFERENCE_TASK_COMPLETIONS = "llm/v1/completions"
+_LLM_INFERENCE_TASK_CHAT = "llm/v1/chat"
+
+_SUPPORTED_LLM_INFERENCE_TASK_TYPES_BY_PIPELINE = {
+    "text-generation": [_LLM_INFERENCE_TASK_COMPLETIONS, _LLM_INFERENCE_TASK_CHAT],
+}
+
+
+def postprocess_output_for_llm_inference_task(
+    data: List[str], output_tensors: List[List[int]], pipeline, model_config, inference_task
+):
+    """
+    Wrap output data with usage information according to the MLflow inference task.
+
+    Example:
+        .. code-block:: python
+            data = ["How to learn Python in 3 weeks?"]
+            output_tensors = [
+                [
+                    1128,
+                    304,
+                    ...,
+                    29879,
+                ]
+            ]
+            output_dicts = postprocess_output_for_llm_inference_task(data, output_tensors, **kwargs)
+
+            assert output_dicts == [
+                {
+                    "text": "1. Start with a beginner's",
+                    "finish_reason": "length",
+                    "usage": {"prompt_tokens": 9, "completion_tokens": 10, "total_tokens": 19},
+                }
+            ]
+
+    Args:
+        data: List of text input prompts.
+        output_tensors: List of output tensors that contain the generated tokens (including
+            the prompt tokens) corresponding to each input prompt.
+        pipeline: The pipeline object used for inference.
+        model_config: The model configuration dictionary used for inference.
+        inference_task: The MLflow inference task.
+
+    Returns:
+        List of dictionaries containing the output text and usage information for each input prompt.
+    """
+    output_dicts = []
+    for input_data, output_tensor in zip(data, output_tensors):
+        output_dict = _get_output_and_usage_from_tensor(
+            input_data, output_tensor, pipeline, model_config, inference_task
+        )
+        output_dicts.append(output_dict)
+
+    return output_dicts
+
+
+def _get_output_and_usage_from_tensor(
+    prompt: str, output_tensor: List[int], pipeline, model_config, inference_task
+):
+    """
+    Decode the output tensor and return the output text and usage information as a dictionary
+    to make the output in OpenAI compatible format.
+    """
+    usage = _get_token_usage(prompt, output_tensor, pipeline, model_config)
+    completions_text = _get_completions_text(prompt, output_tensor, pipeline)
+    finish_reason = _get_finish_reason(
+        usage["total_tokens"], usage["completion_tokens"], model_config
+    )
+
+    output_dict = {"finish_reason": finish_reason, "usage": usage}
+
+    if inference_task == _LLM_INFERENCE_TASK_COMPLETIONS:
+        output_dict["text"] = completions_text
+    elif inference_task == _LLM_INFERENCE_TASK_CHAT:
+        output_dict["messages"] = {"role": "assistant", "content": completions_text}
```

Comment:
Is this role always `assistant`?

Reply:
Yes, the LLM should always return the message in an `assistant` role.